File: Lab4/C#/Vector.cs
using System;

namespace CSharp
{
    public class Vector
    {
        private double _xCoord; // coordinate X
        private double _yCoord; // coordinate Y
        private double _zCoord; // coordinate Z
        private double _length; // length of a Vector

        public Vector() // default constructor, null vector
        {
            this._xCoord = 0;
            this._yCoord = 0;
            this._zCoord = 0;
            this._length = GetLength();
        }

        public Vector(double x, double y, double z) // constructor with parameters(coordinates of vector)
        {
            this._xCoord = x;
            this._yCoord = y;
            this._zCoord = z;
            this._length = GetLength();
        }

        public Vector(Vector vec) // copy of another vector
        {
            this._xCoord = vec._xCoord;
            this._yCoord = vec._yCoord;
            this._zCoord = vec._zCoord;
            this._length = vec.GetLength();
        }

        private double[] GetCoord() // method for getting coordinates of a Vector
        {
            var coord = new double[3] {this._xCoord, this._yCoord, this._zCoord};
            return coord;
        }

        private double GetLength() // method for getting coordinates
        {
            return Math.Sqrt(Math.Pow(this._xCoord, 2) + Math.Pow(this._yCoord, 2) + Math.Pow(this._zCoord, 2));
        }
        
        private double this[int index]
        {
            get
            {
                var coord = this.GetCoord();
                return coord[index];
            }
        }
        
        public void VectorInfo() // printing info about a vector
        {
            Console.WriteLine("Vector`s length: {0}", this._length);
            Console.WriteLine("Vector`s coordinates: ({0}, {1}, {2})\n", this._xCoord, this._yCoord, this._zCoord);
            
        }
        
        public static Vector operator +(Vector vec1, Vector vec2) // vector addition
        {
            var result = new Vector
            {
                _xCoord = vec1._xCoord + vec2._xCoord,
                _yCoord = vec1._yCoord + vec2._yCoord,
                _zCoord = vec1._zCoord + vec2._zCoord,
            };
            result._length = result.GetLength();
            return result;
        }
        public static Vector operator -(Vector vec1, Vector vec2) // vector subtraction
        {
            var result = new Vector
            {
                _xCoord = vec1._xCoord - vec2._xCoord,
                _yCoord = vec1._yCoord - vec2._yCoord,
                _zCoord = vec1._zCoord - vec2._zCoord
            };
            result._length = result.GetLength();
            return result;
        }
        
        public static double operator *(Vector vec1, Vector vec2) //scalar product of vectors
        {
            var sum = .0;
            for (var i = 0; i < 3; i++)
            {
                sum += (vec1[i] * vec2[i]);
            }

            return sum;
        }

        public static Vector operator *(Vector vector, double times) // increase a vector
        {
            vector._xCoord *= times;
            vector._yCoord *= times;
            vector._zCoord *= times;
            vector._length = vector.GetLength();
            return vector;
        } 
        public static Vector operator *(double times, Vector vector) // increase a vector
        {
            return vector * times;
        } 
        
    }
}

File: Lab4/С++/Vector.cpp
#include <iostream>
#include "Vector.hpp"
#include <stdio.h>
#include <math.h>

//return vector`s length
double Vector::GetLength() {
    return sqrt(pow(this->XCoord, 2) + pow(this->YCoord, 2) + pow(this->ZCoord, 2));
}

// return a massive with vector`s coordinates (x, y, z)
double* Vector::GetCoord() {
    double* coord = new double[3] {this->XCoord, this->YCoord, this->ZCoord};
    return coord;
}

// default constructor, null-vector
Vector::Vector() {
    this->XCoord = 0.0;
    this->YCoord = 0.0;
    this->ZCoord = 0.0;
    this->Length = GetLength();
}

// constructor with parameters(coordinates of vector)
Vector::Vector(const double x, const double y, const double z) {
    this->XCoord = x;
    this->YCoord = y;
    this->ZCoord = z;
    this->Length = GetLength();
}

// copy of another vector
Vector::Vector(const Vector &vector){
    this->XCoord = vector.XCoord;
    this->YCoord = vector.YCoord;
    this->ZCoord = vector.ZCoord;
    this->Length = vector.Length;
}

//printing info about vector
void Vector::VectorInfo() {
    std::cout << "Vector`s length: " << this->Length << std::endl;
    printf("Vector`s coordinates: (%f, %f, %f)", GetCoord()[0], GetCoord()[1], GetCoord()[2]);
    std::cout << std::endl;
    std::cout << std::endl;
}

// vector addition
Vector Vector::operator+(Vector &vec){
    Vector newVector;
    newVector.XCoord = this->XCoord + vec.XCoord;
    newVector.YCoord = this->YCoord + vec.YCoord;
    newVector.ZCoord = this->ZCoord + vec.ZCoord;
    newVector.Length = newVector.GetLength();
    return newVector;
}

// vector subtracting
Vector Vector::operator-(Vector &vec) {
    Vector newVector;
    newVector.XCoord = this->XCoord - vec.XCoord;
    newVector.YCoord = this->YCoord - vec.YCoord;
    newVector.ZCoord = this->ZCoord - vec.ZCoord;
    newVector.Length = newVector.GetLength();
    return newVector;
}

// increase vector in X times
Vector Vector::operator*(const double times) {
    Vector newVector;
    newVector.XCoord = this->XCoord * times;
    newVector.YCoord = this->YCoord * times;
    newVector.ZCoord = this->ZCoord * times;
    newVector.Length = newVector.GetLength();
    return newVector;
}

// scalar product
double Vector::operator*(Vector &vec) {
    double sum = 0;
    for (int i = 0; i < 3; i++) {
        sum += (this->GetCoord()[i] * vec.GetCoord()[i]);
    }
    return sum;
}
